File: src/io/innofang/FactoryMethod/CakeFactory.java
package io.innofang.FactoryMethod;

/**
 * Created by InnF on 2017/2/23.
 */
public class CakeFactory extends Factory {
    @Override
    public <T extends Cake> T createProduct(Class<T> clz) {
        Cake cake = null;
        try {
            cake = (Cake) Class.forName(clz.getName()).newInstance();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return (T) cake;
    }
}

File: src/io/innofang/factory_method/example/kotlin/CakeFactory.kt
package io.innofang.factory_method.example.kotlin


/**
 * Created by Inno Fang on 2017/9/1.
 */
class CakeFactory : Factory() {

    override fun <T : Cake> createProduct(clz: Class<T>): T? {
        var cake: Cake? = null
        try {
            cake = Class.forName(clz.name).newInstance() as Cake
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return cake as T?
    }
}